cdc/capture/http_handler.go

```diff
@@ -0,0 +1,544 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package capture
+
+import (
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/gin-gonic/gin"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/ticdc/cdc/kv"
+	"github.com/pingcap/ticdc/cdc/model"
+	"github.com/pingcap/ticdc/cdc/owner"
+	cerror "github.com/pingcap/ticdc/pkg/errors"
+	"github.com/pingcap/ticdc/pkg/httputil"
+	"github.com/pingcap/ticdc/pkg/version"
+	"github.com/tikv/client-go/v2/oracle"
+	"go.etcd.io/etcd/clientv3/concurrency"
+)
+
+const (
+	// APIOpVarChangefeedState is the key of changefeed state in HTTP API
+	APIOpVarChangefeedState = "state"
+	// APIOpVarChangefeedID is the key of changefeed ID in HTTP API
+	APIOpVarChangefeedID = "changefeed_id"
+	// APIOpVarCaptureID is the key of capture ID in HTTP API
+	APIOpVarCaptureID = "capture_id"
+	// APIOpVarTableID is the key of table ID in HTTP API
+	APIOpVarTableID = "table_id"
+)
+
+type httpHandler struct {
+	capture *Capture
+}
+
+// NewHTTPHandler return a httpHandler for OpenAPI
+func NewHTTPHandler(capture *Capture) httpHandler {
+	return httpHandler{
+		capture: capture,
+	}
+}
+
+// ListChangefeed list all changgefeed in cdc cluster
+// @Summary List changefeed
+// @Description list all changefeed in cdc cluster
+// @Tags changefeed
+// @Accept  json
+// @Produce  json
+// @Param   state  query  string  false  "state"
+// @Success 200 {array} model.ChangefeedCommonInfo
+// @Failure 500 {object} model.HTTPError
+// @Router	/api/v1/changefeeds [get]
+func (h *httpHandler) ListChangefeed(c *gin.Context) {
+	state := c.Query(APIOpVarChangefeedState)
+	statuses, err := h.capture.etcdClient.GetAllChangeFeedStatus(c.Request.Context())
+	if err != nil {
+		c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+		return
+	}
+
+	changefeedIDs := make(map[string]struct{}, len(statuses))
+	for cid := range statuses {
+		changefeedIDs[cid] = struct{}{}
+	}
+
+	resps := make([]*model.ChangefeedCommonInfo, 0)
+	for changefeedID := range changefeedIDs {
+		cfInfo, err := h.capture.etcdClient.GetChangeFeedInfo(c.Request.Context(), changefeedID)
+		if err != nil && cerror.ErrChangeFeedNotExists.NotEqual(err) {
+			c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+			return
+		}
+
+		if !httputil.IsFiltered(state, cfInfo.State) {
```

Comment:
The name of the function is confusing. It should be something like `IsNeeded`

Reply:
good